### ISSUE TITLE
Can we set expiration of particular ImageUrl. 

### ISSUE THREAD

**Issue by vaibhavsingla1996** [SDWebImage/SDWebImage#2508]
### New Issue Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / macos / tvos / watchos
 Platform Version        | e.g. 11.0 / 10.13.0 / 11.0 / 4.0
 SDWebImage Version      | e.g. 4.2.0 / 4.1.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 9 / Xcode 8
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Hi,
Can we set the expiration time of cache for particular imageUrls? If yes, then please explain the procedure.
Thanks

**Comment by bpoplauschi** [comment]
@vaibhavsingla1996 thanks for taking the time to create the issue, but please start by filling in the info required in the table above (Issue Info).

The current release 4.x does not support custom handling of each cache type, but since version 5.0 (currently in beta), one can create a custom `SDImageCache` implementation and do whatever inside.

**Comment by dreampiggy** [comment]
Our `SDWebImageDownloader` will flag the URLSession by default, to not use any cache algorithm. Because we already have `SDImageCache`, which will cause a double cache.

Check our wiki : [Custom Cache](https://github.com/rs/SDWebImage/wiki/Advanced-Usage#custom-cache-50). For example, if you want to implement a custom disk cache age for specify url, you can implement `removeExpiredData` method. There are just a pseudocode to show the logic.

```
for cacheFile in cacheFolder {
	if cacheFile.fileName == "target name" {
		if cacheFile.lastModifiedData < 7 days ago {
			fileManager.removeFile(cacheFile)
			return
		}
	}
	super.removeExpiredData()

}
```

**Comment by dreampiggy** [comment]
To say, the cache policy, should really be controlled on the HTTP server side via many [Cache-Control](https://developer.mozilla.org/zh-CN/docs/Web/HTTP/Headers/Cache-Control) header. As a client it's not suitable to modify and provide a custom value again.

So you'd better check your image server at the beginning, to let them specify the correct cache age.

**Comment by dreampiggy** [comment]
If you really want to do this, maybe possbiele. Expect that custom cache implementation, another way, which it's far more harder, but still a possible solution. It's also possible in current 4.x but maybe harder again :)...

Continue to read if you're really want to do this.

---

Except what you can do in custom cache. You can also consider one approach, it's [Custom Download Operation](https://github.com/rs/SDWebImage/wiki/Advanced-Usage#custom-download-operation-40). You need to filter some specify URL, then ignore the `SDImageCache` cache totally (Or you'll have double cache).

Attention this solution need complicated steps and may contains some issue because of URLSession internal process, you'd better use other ways.

#### First Step
You need to disable `SDImageCache`. This can be done in 5.x using `SDWebImageFromLoaderOnly` (stop cache query) && `{SDWebImageContextStoreCacheType : @SDImageCacheTypeNone}` (stop cache storage) options during `sd_setImageWithURL:`

```objective-c
[imageView sd_setImageWithURL:url options:SDWebImageFromLoaderOnly placeholderImage:nil context:@{SDWebImageContextStoreCacheType: @SDImageCacheTypeNone}]
```

#### Second Step
Next, you need to restore the [NSURLRequest.cachePolicy](https://developer.apple.com/documentation/foundation/nsurlrequest/cachepolicy). Because by default we use `NSURLRequestReloadIgnoringLocalCacheData`, which will cause that URLSession delegate method not been called at all. This step can be achieved using []().

```objective-c
SDWebImageDownloaderRequestModifierBlock requestModifier = ^NSURLRequest *(NSURLRequest * request) {
    NSMutableURLRequest *mutableRequest = [request mutableCopy];
    if ([request.URL.absoluteString isEqualToString:@"http://foo/bar"]) {
        mutableRequest.cachePolicy = NSURLRequestUseProtocolCachePolicy;
    }
    return [mutableRequest copy];
};
```

#### Third Step
Last, you need to subclass `SDWebImageDownloadOperation` and override that our running method `start`. I provide a demo code base on the logic (but not test whether it's possible to work or not), you can check about it.


```objective-c
- (void)start {
    if ([self.request.URL.absoluteString isEqualToString:@"http://foo/bar"]) {
        NSCachedURLResponse *cachedResponse = [URLCache cachedResponseForRequest:self.request];
        if (cachedResponse) {
            NSString *lastDateString = (((NSHTTPURLResponse *)cachedResponse.response).allHeaderFields[@"Date"]);
            NSDateFormatter *dateFormatter = [NSDateFormatter new];
            dateFormatter.dateFormat = @"EEE',' dd' 'MMM' 'yyyy HH':'mm':'ss zzz";
            NSDate *lastDate = [dateFormatter dateFromString:lastDateString];
            NSDate *expirationDate = [NSDate dateWithTimeIntervalSinceNow:-7 * 3600 * 24]; // 7 days ago
            if ([[lastDate laterDate:expirationDate] isEqualToDate:expirationDate]) {
                // expire and remove
                [URLCache removeCachedResponseForRequest:self.request];
            }
        }
    }
    [super start];
}
```

Then connect all of this done together.
